### PR TITLE
track adt variant and field names do not conflict

### PR DESCRIPTION
**Pull request by shua**
The intent of this change is to add a check for adts that enum variant names are unique, and that struct and enum field names are unique (within a given enum variant).

### CODE REVIEW

File: crates/formality-check/src/adts.rs [hunk 1]
```diff
@@ -15,10 +18,17 @@ impl super::Check<'_> {
 
         self.prove_where_clauses_well_formed(&env, &where_clauses, &where_clauses)?;
 
-        // FIXME: check names are unique or integers from 0..n
-
-        for Variant { name: _, fields } in &variants {
-            for Field { name: _, ty } in fields {
+        // names is used to check that there are no name conflicts
+        let mut names = HashSet::new();
```
Comment (lcnr) [inline crates/formality-check/src/adts.rs:13]: can you move the name check above any trait solving? i.e. right after the `instantiate_universally`?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
this looks good modulo @lcnr's suggestion

File: crates/formality-check/src/adts.rs (hunk 1 above)
Comment (shua) [reply crates/formality-check/src/adts.rs:13]: I needed some way to get the term out of the `Binder`. I guess it doesn't matter how it's instantiated for the purposes of unique naming, so I chose `Binder::open()` but maybe there's a simpler way with eg `Binder::instantiate`?

Comment (nikomatsakis) [reply crates/formality-check/src/adts.rs:13]: you can use the `Binder::peek` method, @shua

**Review by nikomatsakis — APPROVED** [review]

File: crates/formality-check/src/adts.rs [hunk 2]
```diff
@@ -6,6 +9,20 @@ impl super::Check<'_> {
     pub(super) fn check_adt(&self, adt: &Adt) -> Fallible<()> {
         let Adt { id: _, binder } = adt;
 
+        // names is used to check that there are no name conflicts
+        let mut names = HashSet::new();
+        for Variant { name, fields } in &adt.binder.open().1.variants {
```
Comment (nikomatsakis) [inline crates/formality-check/src/adts.rs]:
```suggestion
        for Variant { name, fields } in &adt.binder.skip().variants {
```

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: crates/formality-check/src/adts.rs (hunk 2 above)
Comment (shua) [reply crates/formality-check/src/adts.rs]: I saw you merged this change, but `skip` doesn't exist on binder so I changed it to the previously mentioned `peek` method instead.

Comment (nikomatsakis) [reply crates/formality-check/src/adts.rs]: Oh, ooops, thanks!

**Review by nikomatsakis — APPROVED** [review]

File: src/test/mod.rs [hunk 3]
```diff
@@ -136,6 +136,41 @@ fn basic_where_clauses_fail() {
...
+        [
+            crate Foo {
+                enum Bar {
+                    Baz{},
+                    Baz{},
+                }
+            }
+        ]
+
+        [ /* TODO */ ]
```
Comment (nikomatsakis) [inline src/test/mod.rs]: add suggested messages in here

File: src/test/mod.rs [hunk 4]
```diff
@@ -136,6 +136,41 @@ fn basic_where_clauses_fail() {
...
+        [
+            crate Foo {
+                struct Bar {
+                    baz: (),
+                    baz: (),
+                }
+            }
+        ]
+
+        [ /* TODO */ ]
```
Comment (nikomatsakis) [inline src/test/mod.rs]: and here

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Almost there!

File: src/test/mod.rs (hunk 3 above)
Comment (shua) [reply src/test/mod.rs]: Ah, I copied this from pre-existing tests in this file, so I had no idea what to put here, but scrolled down and found some other examples. This is supposed to be some summary of what's wrong with the code?

Comment (shua) [reply src/test/mod.rs]: ah wait, maybe instead it's some relevant substring in the error output that shouldn't change even when you update the expect output? (guessing from `assert_err!` naming that tt `must_have` and the other `expect`)

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**